### PR TITLE
git-cli task

### PR DESCRIPTION
**Pull request by Aneesh-M-Bhat**

### CODE REVIEW

File: scripts/git-cli-prepare.sh
Comment (otaviof) [inline scripts/git-cli-prepare.sh:1]: Basically the same actions needed for prepare are [present on the this script already](https://github.com/openshift-pipelines/task-git/blob/6c85da21eae3f29bd80869ff5d97006d7e0c254e/scripts/prepare.sh), only the sequence chosen is different. Let's use a single "prepare" script for both?

File: scripts/git-cli.sh
Comment (otaviof) [inline scripts/git-cli.sh:1]: I think we should consider a single "git run" script for both "git clone" and "git cli" Tasks. Configuration like HTTP(S) proxy, `git config` and others become a common pre-flight activity for both Tasks.

https://github.com/openshift-pipelines/task-git/blob/6c85da21eae3f29bd80869ff5d97006d7e0c254e/scripts/git-clone.sh

File: templates/environment.tpl
Comment (otaviof) [inline templates/environment.tpl:1]: Could you please have these modifications on [existing `environment-variables.tpl`](https://github.com/openshift-pipelines/task-git/blob/6c85da21eae3f29bd80869ff5d97006d7e0c254e/templates/environment-variables.tpl)? Also the changes for having `git-clone` Task working with a informed slice of environment variables, like for instance what's in [place for `task-containers` repository](https://github.com/openshift-pipelines/task-containers/blob/3808765d18f17bdaf8a3a72d7df2cddfac2a45be/templates/task-buildah.yaml#L49-L67).

File: templates/load-scripts.tpl
Comment (otaviof) [inline templates/load-scripts.tpl:1]: Could you please adopt the same on `git-clone` Task? It's good to have this logic on this repository as well 👍

File: templates/task-git-cli.yaml [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    - name: DELETE_EXISTING
+      type: string
+      default: "true"
+      description: |
+        Clean out the contents of the `output` Workspace before cloning the
+        repository, if data exists.
+
+  results:
+    - name: COMMIT
+      description: The precise commit SHA after the git operation.
```
Comment (otaviof) [inline templates/task-git-cli.yaml:98]: Most of the things employed on this Task [is implemented on `task.yaml`](https://github.com/openshift-pipelines/task-git/blob/6c85da21eae3f29bd80869ff5d97006d7e0c254e/templates/task.yaml#L16-L132). I think we should follow the same design taken on `task-s2i.yaml` ([1](https://github.com/openshift-pipelines/task-containers/blob/3808765d18f17bdaf8a3a72d7df2cddfac2a45be/templates/spec-s2i.tpl) and [2](https://github.com/openshift-pipelines/task-containers/blob/3808765d18f17bdaf8a3a72d7df2cddfac2a45be/templates/task-s2i-go.yaml)), thus the user experience is the same for the Git related Tasks in this repository.

File: test/e2e/e2e.bats
Comment (otaviof) [inline test/e2e/e2e.bats:1]: Would be great to use `git mv` instead, please. Thus the commit file history won't be lost 🙏

File: test/e2e/resources/pipeline-git-cli.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    #         script: |
+    #           #!/usr/bin/env sh
+    #           cd $(workspaces.source.path)
+    #           receivedCommit=$(git rev-parse HEAD)
+    #           if [ $receivedCommit != $(params.expected-commit) ]; then
+    #             echo "Expected commit $(params.expected-commit) but received $receivedCommit."
+    #             exit 1
+    #           else
+    #             echo "Received commit $receivedCommit as expected."
+    #           fi
```
Comment (otaviof) [inline test/e2e/resources/pipeline-git-cli.yaml:67]: Please avoid leaving commented lines 😇

**Review by otaviof — CHANGES_REQUESTED** [review]
Please consider the review comments, and also add a PR description as well 🙏

**Comment by otaviof** [comment]
I wonder why GHA haven't run on this PR, WDYT @vdemeester? 🤔 At least the e2e tests should have run for the PR, but nothing.

File: scripts/git-run.sh
Comment (otaviof) [inline scripts/git-run.sh:1]: Good move merging all the `git` related things on a single script 👍

File: scripts/git-run.sh [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    exec git-init \
+        -url="${PARAMS_URL}" \
+        -revision="${PARAMS_REVISION}" \
+        -refspec="${PARAMS_REFSPEC}" \
+        -path="${checkout_dir}" \
+        -sslVerify="${PARAMS_SSL_VERIFY}" \
+        -submodules="${PARAMS_SUBMODULES}" \
+        -depth="${PARAMS_DEPTH}" \
+        -sparseCheckoutDirectories="${PARAMS_SPARSE_CHECKOUT_DIRECTORIES}"
+fi
```
Comment (otaviof) [inline scripts/git-run.sh:40]: Given the script either run `git clone` (`git-init`) or an arbitrary "git script", this block should become a `if/else`.

File: scripts/common.sh [hunk 4]
```diff
@@ -46,15 +49,14 @@ phase() {
 
 # Inspect the environment variables to assert the minimum configuration is informed.
 assert_required_configuration_or_fail() {
-    [[ -z "${PARAMS_URL}" ]] &&
-        fail "Parameter URL is not set!"
+    ([[ -z "${PARAMS_URL}" ]] && [[ -z "${PARAMS_GIT_SCRIPT}" ]]) &&
+        fail "Parameter URL or SCRIPT is not set!"
```
Comment (otaviof) [inline scripts/common.sh:53]: Additionally, when both params are informed it should `fail` as well.

File: scripts/prepare.sh [hunk 5]
```diff
@@ -30,9 +35,44 @@ if [[ "${WORKSPACES_SSH_DIRECTORY_BOUND}" == "true" ]]; then
...
+
+phase "Setting up HTTP_PROXY='${PARAMS_HTTP_PROXY}'"
+[[ -n "${PARAMS_HTTP_PROXY}" ]] && export HTTP_PROXY="${PARAMS_HTTP_PROXY}"
+
+phase "Settting up HTTPS_PROXY='${PARAMS_HTTPS_PROXY}'"
+[[ -n "${PARAMS_HTTPS_PROXY}" ]] && export HTTPS_PROXY="${PARAMS_HTTPS_PROXY}"
+
+phase "Setting up NO_PROXY='${PARAMS_NO_PROXY}'"
+[[ -n "${PARAMS_NO_PROXY}" ]] && export NO_PROXY="${PARAMS_NO_PROXY}"
+
```
Comment (otaviof) [inline scripts/prepare.sh:77]: Given the Task is split into three steps ([`prepare`, `git-clone` and `report`](https://github.com/openshift-pipelines/task-git/blob/6c85da21eae3f29bd80869ff5d97006d7e0c254e/templates/task.yaml#L166-L192)), the exported variables won't be passed from `prepare` to `git-clone`. And, that's why the `export` settings were bundled on `git-clone.sh` before.

I think the best course of action is to make sure any `export` variable is performed right before `git-init` (or an arbitrary `git` script).

File: templates/common-steps.tpl [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+- name: prepare
+  image: {{ .Values.images.gitInit }}
+  workingDir: $(workspaces.source.path)
+  command:
+    - /scripts/prepare.sh
+  volumeMounts:
+    - name: scripts-dir
+      mountPath: /scripts
+    - name: user-home
+      mountPath: "$(params.USER_HOME)"
```
Comment (otaviof) [inline templates/common-steps.tpl:11]: I think the `prepare` step is shared between the two Tasks, `git-clone` and `git-cli`.

File: templates/environment.tpl
Comment (otaviof) [inline templates/environment.tpl:1]: Please make sure [`environment-variables.tpl`](https://github.com/openshift-pipelines/task-git/blob/6c85da21eae3f29bd80869ff5d97006d7e0c254e/templates/environment-variables.tpl) is included on the scope of this pull-request.

File: test/e2e/e2e.bats
Comment (otaviof) [inline test/e2e/e2e.bats:1]: Please use `git mv` to keep the commit history of `e2e.bats` moved to `e2e-git-clone.bats` 🙏

File: test/e2e/resources/task-git-cli.yaml
Comment (otaviof) [inline test/e2e/resources/task-git-cli.yaml:1]: I think you can re-use existing PVC from `git clone`, it's easier to keep a single PVC for testing purposes.

**Review by otaviof — CHANGES_REQUESTED** [review]
Good progress on this 👏 I left another round of comments, and I think we should take in consideration `task-containers` S2I approach, where a [single template](https://github.com/openshift-pipelines/task-containers/blob/3808765d18f17bdaf8a3a72d7df2cddfac2a45be/templates/spec-s2i.tpl) renders all the "git" Tasks flavours, so we can keep the logic into a single place ([but rendering two Tasks](https://github.com/openshift-pipelines/task-containers/blob/3808765d18f17bdaf8a3a72d7df2cddfac2a45be/templates/task-s2i-go.yaml)).

File: templates/common-steps.tpl (hunk 6 above)
Comment (Aneesh-M-Bhat) [reply templates/common-steps.tpl:11]: I didnt get you, prepare step is shared thats true, so im keeping it in the steps template, is there any other way you are trying to say?

File: templates/environment.tpl
Comment (Aneesh-M-Bhat) [reply templates/environment.tpl:1]: I didnt get you, I have renamed it & removed uncommon params from it. Is there something wrong with the approach?

**Comment by vdemeester** [comment]
@Aneesh-M-Bhat needs a rebase 🙏🏼

File: scripts/common.sh [hunk 7]
```diff
@@ -46,15 +49,17 @@ phase() {
...
```
Comment (otaviof) [inline scripts/common.sh:53]: Good move 👍 I think the erro message should mention something like "Only Param URL or SCRIPT must be set!". And also, please consider the following syntax:

```bash
    [[ -z "${PARAMS_URL}" -a -z "${PARAMS_GIT_SCRIPT}" ]] &&
        fail "Only Param URL or SCRIPT must be set!"
```

File: scripts/prepare.sh [hunk 8]
```diff
@@ -30,9 +35,17 @@ if [[ "${WORKSPACES_SSH_DIRECTORY_BOUND}" == "true" ]]; then
...
 fi
 
-if [[ "${PARAMS_DELETE_EXISTING}" == "true" ]]; then
-	phase "Deleting all contents of checkout-dir '${checkout_dir}'"
-	clean_dir ${checkout_dir} || true
+# Setting up the config for the git.
+
+if [ -n "${PARAMS_GIT_USER_EMAIL}" ] ; then
+	echo "${[ -n "${PARAMS_GIT_USER_EMAIL}"]}"
+    #phase "Setting global email for git ${PARAMS_GIT_USER_EMAIL}"
```
Comment (otaviof) [inline scripts/prepare.sh:42]: nit: Uncomment the `phase` statement.

File: scripts/prepare.sh [hunk 9]
```diff
@@ -30,9 +35,17 @@ if [[ "${WORKSPACES_SSH_DIRECTORY_BOUND}" == "true" ]]; then
...
+    git config --global user.email "${PARAMS_GIT_USER_EMAIL}"
+fi
+
+if [ -n "${PARAMS_GIT_USER_NAME}" ] ; then
+    #phase "Setting global username for git ${PARAMS_GIT_USER_NAME}"
```
Comment (otaviof) [inline scripts/prepare.sh:47]: nit: Uncomment the `phase` statement.

File: test/e2e/e2e.bats
Comment (otaviof) [inline test/e2e/e2e.bats:1]: Please use `git mv` here too, from `e2e.bats` to `e2e-git-clone.bats`.

**Review by otaviof — COMMENTED** [review]
Really good progress, just a couple of minor/nit comments 🎉

File: scripts/common.sh (hunk 7 above)
Comment (Aneesh-M-Bhat) [reply scripts/common.sh:53]: instead of -a, i had to &&

File: templates/common-workspaces.tpl [hunk 10]
```diff
@@ -0,0 +1,33 @@
+{{- /*
+
+  Contains the results produced by buildah, should be part of a Tasks ".spec.results[]".
+
+*/ -}}
+{{- define "common_workspaces" -}}
```
Comment (vdemeester) [inline templates/common-workspaces.tpl:6]: For the `git-clone` `Task`, the "main" workspace needs to be called `output` in order to be backward compatible.

**Review by vdemeester — COMMENTED** [review]
We need to ensure we do backward compatible changes with the current ClusterTask shipped in OpenShift Pipelines (at least in the first version, and while there is still ClusterTask).

File: test/e2e/e2e.bats
Comment (Aneesh-M-Bhat) [reply test/e2e/e2e.bats:1]: yep, had to do changes on initial file, then push it.

File: templates/common-workspaces.tpl (hunk 10 above)
Comment (Aneesh-M-Bhat) [reply templates/common-workspaces.tpl:6]: cool, will do the changes then.

**Comment by vdemeester** [comment]
```
 # ./test/helper/helper.sh: line 27: declare: tmpl_file: readonly variable 
```

**Review by vdemeester — APPROVED** [review]
/lgtm

**Pull request merged**